packages/realm-react/src/useObject.tsx

```diff
@@ -55,24 +55,96 @@ export function createUseObject(useRealm: () => Realm) {
+
     // Wrap the cachedObject in useMemo, so we only replace it with a new instance if `primaryKey` or `type` change
     const { object, tearDown } = useMemo(
       // TODO: There will be an upcoming breaking change that makes objectForPrimaryKey return null
       // When this is implemented, remove `?? null`
-      () =>
-        createCachedObject({
-          object: realm.objectForPrimaryKey<unknown>(type, primaryKey) ?? null,
-          realm,
-          updateCallback: forceRerender,
-        }),
-      [type, realm, primaryKey],
+      () => {
+        // This should never happen, but if it does, we want to return a null result
+        if (!cachedObjectRef.current) {
+          return { object: null, tearDown: () => undefined };
+        }
+
+        // Re-instantiate the cachedObject if the primaryKey has changed or the originalObject has gone from null to not null
+        if (
+          !arePrimaryKeysIdentical(primaryKey, primaryKeyRef.current) ||
+          (originalObjectRef.current === null && originalObject !== null)
+        ) {
+          cachedObjectRef.current = createCachedObject({
+            object: originalObject ?? null,
+            realm,
+            updateCallback: forceRerender,
+          });
+          originalObjectRef.current = originalObject;
+        }
+        return cachedObjectRef.current;
+      },
+      [realm, originalObject, primaryKey],
     );
 
     // Invoke the tearDown of the cachedObject when useObject is unmounted
     useEffect(() => {
       return tearDown;
     }, [tearDown]);
 
+    // If the object doesn't exist, listen for insertions to the collection and force a rerender if the inserted object has the correct primary key
+    useEffect(() => {
+      const collectionListener: Realm.CollectionChangeCallback<T & Realm.Object> = (collection, changes) => {
+        const primaryKeyProperty = collection?.[0]?.objectSchema()?.primaryKey;
+        for (const index of changes.insertions) {
+          const object = collection[index];
+          if (primaryKeyProperty) {
+            //@ts-expect-error - if the primaryKeyProperty exists, then it is indexable. However, we don't allow it when we don't actually know the type of the object
+            const insertedPrimaryKey = object[primaryKeyProperty];
+            if (arePrimaryKeysIdentical(insertedPrimaryKey, primaryKeyRef.current)) {
+              forceRerender();
+              collection.removeListener(collectionListener);
+              break;
+            }
+          }
+        }
+      };
+
+      if (!originalObjectRef.current) {
+        const collection = realm.objects(type);
+        collection.addListener(collectionListener);
+      }
+
+      return () => {
+        if (!originalObjectRef.current) {
+          // If the app is closing, the realm will be closed and the listener does not need to be removed
+          if (!realm.isClosed) {
+            const collection = realm.objects(type);
```

Comment:
I'm actually not sure what this will do. Are you sure adding a listener to one accessor object and removing it from another will actually stop notifications? I suspect you'd need to hold a reference to the original collection accessor object for this removal to work as expected.

Reply:
That's possible that this is the case and I will look into it.  

This specific case is not being tested, but should probably be proved in our unit tests somewhere.  

It is difficult to test such specifics within hooks, without creating extra an extra accessor layer on top of realm.  One could argue that we do that, but I would argue that we already have a test harness to test such specifics.

Reply:
Went over this in the morning.  At some point it would be good to find time to refactor code to be more testable.  The potential listener issues has now been fixed.